Sally7/S7Connection.cs

```diff
@@ -167,11 +203,26 @@ public async Task WriteAsync(IDataItem[] dataItems, CancellationToken cancellati
             Memory<byte> mem = mo.Memory;
             int length = S7ConnectionHelpers.BuildWriteRequest(mem.Span, dataItems);
 
-            // The response will only be written after the request has been sent. At that point we no longer
-            // care about the request contents, so we use a single buffer only.
-            Memory<byte> response = await executor.PerformRequest(mem.Slice(0, length), mem, cancellationToken).ConfigureAwait(false);
+            using var linkedCts = CreateRequestTimeoutCancellationTokenSource(cancellationToken);
+            var linkedToken = linkedCts.Token;
 
-            S7ConnectionHelpers.ParseWriteResponse(response.Span, dataItems);
+            try
+            {
+                // The response will only be written after the request has been sent. At that point we no longer
+                // care about the request contents, so we use a single buffer only.
+                Memory<byte> response = await executor.PerformRequest(mem.Slice(0, length), mem, linkedToken)
+                    .ConfigureAwait(false);
+
+                S7ConnectionHelpers.ParseWriteResponse(response.Span, dataItems);
+            }
+            catch (OperationCanceledException) when (linkedToken.IsCancellationRequested)
```

Comment:
Can you explain why it is necessary in the first place to re-throw with the appropriate token?

I would expect in most cases if a OperationCancelledException is thrown by the downstream code having called `linkedToken.ThrowIfCancellationRequested();` that this would already get the exception for whichever of the 2 linked tokens was cancelled.

Or is this solely about ordering the user cancellationToken's cancellation state as higher than the Timeout Exception?

As far as I remember, cancellation in .NET was a best effort thing anyway, and so I am wondering why I as a caller would be able to rely on getting a cancellation exception from my token at all costs - instead of just being able live with it, no matter whether the cancellation came from my own token or from the configured timeout.

Reply:
You'd actually get an exception with `linkedToken` as `CancellationToken`, which makes it impossible to detect if the cancellation happened due to the token being passed in begin cancelled. I guess for day to day usage it's perhaps irrelevant where the cancellation came from, probably especially in the context of Sally7 where it's really shallow. However, when things go south and you start checking the `CancellationToken` of the exception it starts to matter a lot more.

I think I made good effort on the usability by throwing `TimeoutException` when applicable, but to me it appears this is the right way to do it, even if that takes another 2 lines of code.

Reply:
> As far as I remember, cancellation in .NET was a best effort thing anyway, and so I am wondering why I as a caller would be able to rely on getting a cancellation exception from my token at all costs - instead of just being able live with it, no matter whether the cancellation came from my own token or from the configured timeout.

> I still don't think this will hint you the correct token that caused the cancellation all the time. If both cancellation tokens have been cancelled at the time you are in the catch block, you can't technically tell which one came first. Your code just "arbitrarily" (but that might be a bit of strong phrasing here) prefers to throw from the users cancellation token first if it is canceled.

Yes, cancellation is a best effort thing. because often you want to cancel things you don't control, so instead you cancel further processing of what would happen next. You're right about the race condition with regards to user cancellation and timeout, especially if the passed in token would have an identical timeout as the `RequestTimeout`. Assuming both tokens cancel at the same time there's no way to know which one was first, but I can't imagine why you would still care at that point. And if you would care, it would be fixable by changing the timeout. I have considered not adding the `RequestTimeout` and deferring the responsibility to consumers of the library, but I think this makes it easier for consumers while not polluting the library too much. After all I went from 2 timeout properties to 1 and it actually does something this time around...

Reply:
Thanks to your reply I also figured that I should probably close the connection if `Send` is cancelled. We reuse job ID's so if you cancel send, it ends up in the PLC and it replies, then the next job might have taken that ID. Will get that handled as well...